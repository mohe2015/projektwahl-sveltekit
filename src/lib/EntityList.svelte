<!--
SPDX-License-Identifier: AGPL-3.0-or-later
SPDX-FileCopyrightText: 2021 Moritz Hedtke <Moritz.Hedtke@t-online.de>
-->
<script lang="typescript">
	import { query as query2 } from './writable_url';
	import type { Writable } from 'svelte/store';
	import type { BaseEntityType, BaseQueryType, EntityResponseBody } from './entites';
	import { writable } from 'svelte/store';
	import { goto, invalidate } from '$app/navigation';
	import { page } from '$app/stores';

	export let initialQuery: BaseQueryType; // TODO FIXME we need generic components
	export let title: string;
	export let createUrl: string;
	export let response: EntityResponseBody;
	export let fullInvalidationUrl: string;

	export let query: Writable<BaseQueryType> = query2<BaseQueryType>(initialQuery);

	// TODO FIXME A/B testing for sorting (whether to priority first or last chosen option)
	// you wanna sort for type then name

	// TODO FIXME should sorting order change reset to page 1
	// TODO FIXMe should filter change reset to page 1?

	let paginationDirection: Writable<'forwards' | 'backwards' | null> = writable(null);
	let paginationCursor: Writable<BaseEntityType | null> = writable(null);

	export const headerClick = (sortType: string): void => {
		let value = $query['sorting[]'];
		let oldElementIndex = value.findIndex((e) => e.startsWith(sortType + ':'));
		let oldElement = value.splice(oldElementIndex, 1)[0];

		let newElement: string;
		switch (oldElement.split(':')[1]) {
			case 'down-up':
				newElement = 'ASC';
				break;
			case 'ASC':
				newElement = 'DESC';
				break;
			default:
				newElement = 'down-up';
		}

		$query['sorting[]'] = [...value, oldElement.split(':')[0] + ':' + newElement];
	};

	export async function refresh() {
		console.log('invalidate');
		await invalidate(fullInvalidationUrl);
	}

	export const currentSortValue = (sorting: string[], sortingType: string): string => {
		return sorting.find((e) => e.startsWith(sortingType + ':'))?.split(':')[1] ?? '';
	};
</script>

<svelte:head>
	<title>{title}</title>
</svelte:head>

<h1 class="text-center">{title}</h1>

<div class="row justify-content-between">
	<div class="col-auto">
		<a class="btn btn-primary" href={createUrl} role="button">{title} erstellen</a>
		<slot name="buttons" />
	</div>

	<div class="col-3">
		<!-- svelte-ignore a11y-no-onchange -->
		<select
			bind:value={$query.pagination_limit}
			class="form-select"
			aria-label="Default select example"
		>
			<option value="10">10 pro Seite</option>
			<option value="25">25 pro Seite</option>
			<option value="50">50 pro Seite</option>
			<option value="100">100 pro Seite</option>
		</select>
	</div>
</div>

<table class="table">
	<slot name="filter" {headerClick} {currentSortValue} {query} />
	<slot name="response" {response} />
</table>

<nav aria-label="Navigation der Nutzerliste">
	<ul class="pagination justify-content-center">
		<li class="page-item {response.previousCursor ? '' : 'disabled'}">
			<a
				on:click|preventDefault={async () => {
					await goto($page.path + '?' + $page.query, {
						state: {
							paginationCursor: response.previousCursor,
							paginationDirection: 'backwards'
						}
					});
					await refresh(); // TODO FIXME backwards pagination should also invalidate
				}}
				class="page-link"
				href="/"
				aria-label="Vorherige Seite"
				tabindex={response.previousCursor ? undefined : -1}
				aria-disabled={!response.previousCursor}
			>
				<span aria-hidden="true">&laquo;</span>
			</a>
		</li>
		<li class="page-item {response.nextCursor ? '' : 'disabled'}">
			<a
				on:click|preventDefault={async () => {
					await goto($page.path + '?' + $page.query, {
						state: {
							paginationCursor: response.nextCursor,
							paginationDirection: 'forwards'
						}
					});
					await refresh();
				}}
				class="page-link"
				href="/"
				aria-label="NÃ¤chste Seite"
				tabindex={response.nextCursor ? undefined : -1}
				aria-disabled={!response.nextCursor}
			>
				<span aria-hidden="true">&raquo;</span>
			</a>
		</li>
	</ul>
</nav>
