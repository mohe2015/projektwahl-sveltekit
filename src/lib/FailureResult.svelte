<!--
SPDX-License-Identifier: AGPL-3.0-or-later
SPDX-FileCopyrightText: 2021 Moritz Hedtke <Moritz.Hedtke@t-online.de>
-->
<script lang="ts">
import { isErr, Result } from "./result";


	export let result: Result<unknown, { [key: string]: string }>;
</script>

{#if isErr(result)}
	<div class="alert alert-danger" role="alert">
		Fehler:<br />
		{#each Object.entries(result.failure) as [attribute, message]}
			{attribute}: {message}<br />
		{/each}
	</div>
{/if}
