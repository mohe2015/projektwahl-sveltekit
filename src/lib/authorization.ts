// SPDX-License-Identifier: AGPL-3.0-or-later
// SPDX-FileCopyrightText: 2021 Moritz Hedtke <Moritz.Hedtke@t-online.de>
import type { JSONValue } from '@sveltejs/kit/types/helper';
import type { ServerRequest } from '@sveltejs/kit/types/hooks';
import type { MyLocals } from 'src/hooks';
import type { Existing, RawUserType } from './types';

export class HTTPError extends Error {
	status: number;
	statusText: string;

	constructor(status: number, statusText: string) {
		super(`HTTP-Fehler ${status} ${statusText}`);
		this.name = 'HTTPError';
		this.status = status;
		this.statusText = statusText;
	}
}

export const allowUserType = (
	request: ServerRequest<MyLocals, unknown>,
	allowedTypes: ('voter' | 'helper' | 'admin')[]
): Existing<RawUserType> => {
	if (!request.locals.user || !(allowedTypes as string[]).includes(request.locals.user.type)) {
		throw new HTTPError(403, 'Forbidden');
	}
	return request.locals.user;
};

export const allowAnyone = (_request: ServerRequest<MyLocals, unknown>): void => {
	// do nothing.
};

export type PermissionType = {
	view: (user: Existing<RawUserType> | null, entity: JSONValue) => boolean;
	edit: (user: Existing<RawUserType> | null, entity: JSONValue) => boolean;
};

export type PermissionsType<T> = {
	[index in keyof T]: PermissionType;
};

export const checkPermissions = <T>(
	permissions: PermissionsType<T>,
	user: Existing<RawUserType> | null,
	body: JSONValue,
	mode: 'view' | 'edit'
): T => {
	if (typeof body !== 'object' || Array.isArray(body) || body == null) {
		throw new HTTPError(401, `invalid type`);
	}
	const sanitizedValue: {
		[key: string]: string | number | boolean | null;
	} = {};
	for (const [key, checker] of Object.entries<PermissionType>(permissions)) {
		if (body[key] === undefined) continue;

		if (mode === 'edit') {
			if (!checker.edit(user, body)) {
				throw new HTTPError(401, `not allowed to change ${key}`);
			}
		}
		if (mode === 'view') {
			if (!checker.view(user, body)) {
				throw new HTTPError(401, `not allowed to view ${key}`);
			}
		}
		const val = body[key];
		if (val && (typeof val === 'object' || Array.isArray(val))) {
			throw new HTTPError(401, `invalid type at ${key}`);
		}
		sanitizedValue[key] = val;
	}
	for (const [key, _value] of Object.entries(body)) {
		if (!(key in permissions)) {
			throw new HTTPError(401, `additional ignored field ${key}`);
		}
	}
	// TODO FIXME check types of fields otherwise this is not correct
	return sanitizedValue as unknown as T;
};
