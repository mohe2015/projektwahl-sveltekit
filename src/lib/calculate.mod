# SPDX-FileCopyrightText: 2021 Moritz Hedtke <Moritz.Hedtke@t-online.de>
# SPDX-License-Identifier: AGPL-3.0-or-later
# http://gusek.sourceforge.net/gmpl.pdf
# https://ampl.com/resources/the-ampl-book/chapter-downloads/
# glpsol --math src/lib/calculate.mod --wlp test --nopresol --output a
# --data test.dat

# WARNING: if a linear conditional evaluates to nothing it will add a fake 0 * somerandomvariable it seems as the lp format probably doesnt support an empty format

#If you encounter problems with your MathProg model, you can investigate further by specifying the GLPSOL options --nopresol to disable the LP presolver and --output filename.out to write the final solution to a text file. 

set U; # users

set P; # projects

set PA := {'min_participants', 'max_participants'};

# choices (if the user voted something the others get -1 then?)
param choices{u in U, p in P} integer default -1;

param projects{p in P, pa in PA} integer;

param project_leaders{u in U} symbolic; # -1 for not a project leader

var user_in_project{u in U, p in P} binary;

var project_not_exists{p in P} binary;

var user_is_project_leader{u in U} binary;

# TODO FIXME user not in project they are project leader in

maximize total_cost: sum {u in U, p in P} if choices[u,p] != -1 then choices[u,p] * user_in_project[u,p];

subject to notinprojectyoudidntvote{u in U, p in P}:
    if choices[u,p] == -1 then user_in_project[u,p] = 0;

subject to no_project_leader{u in U}: if project_leaders[u] == 'null' then user_is_project_leader[u] else 0 = 0;

subject to either_project_leader_or_project_not_exists{u in U}:
    if project_leaders[u] != 'null' then user_is_project_leader[u] + project_not_exists[project_leaders[u]] else 1 = 1;

subject to onlyinoneproject{u in U}: (sum {p in P} user_in_project[u,p]) + user_is_project_leader[u] = 1;

subject to project_min_size{p in P}: (sum {u in U} user_in_project[u,p]) + projects[p,'min_participants'] * project_not_exists[p] >= projects[p,'min_participants'];
subject to project_max_size{p in P}: (sum {u in U} user_in_project[u,p]) + projects[p,'max_participants'] * project_not_exists[p] <= projects[p,'max_participants'];

data;

set P := project0 project1 project2;

set U := user0 user1 user2;

param project_leaders [user0] null [user1] null [user2] project0;

param projects : min_participants max_participants :=
project0         1                1
project1         1                2
project2         1                5                ;

param choices [user0, project0] 1 [user1, project1] 1 [user2, project2] 1;

end;
