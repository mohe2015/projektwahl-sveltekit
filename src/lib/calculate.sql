-- SPDX-License-Identifier: AGPL-3.0-or-later
-- SPDX-FileCopyrightText: 2021 Moritz Hedtke <Moritz.Hedtke@t-online.de>
SELECT t.user_id, t.project_id, t.rank FROM users, LATERAL
    (WITH c AS (SELECT * FROM choices WHERE choices.user_id = users.id LIMIT 6), c_count AS (SELECT COUNT(*) AS count FROM c)
        SELECT user_id, project_id, rank FROM c WHERE (SELECT count FROM c_count) = 5
        UNION ALL
        -- TOOD FIXME is this part even needed and are the AND at the end correct (shouldn't they be ORs and aren't they checked by triggers already?)
        SELECT users.id as user_id, projects.id AS project_id, -1 as rank FROM projects WHERE (SELECT count FROM c_count) != 5 AND users.age >= projects.min_age AND users.age <= projects.max_age AND users.project_leader_id IS DISTINCT FROM projects.id
    )
    t;
-- 33 ms