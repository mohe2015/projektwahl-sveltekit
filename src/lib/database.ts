// SPDX-License-Identifier: AGPL-3.0-or-later
// SPDX-FileCopyrightText: 2021 Moritz Hedtke <Moritz.Hedtke@t-online.de>

import postgres from 'postgres';
import dotenv from 'dotenv';

// TODO FIXME hack because VITE doesn't load all env vars
dotenv.config();

export const sql = postgres(process.env['DATABASE_URL']!, {
	debug: (connection, query, params) => {
		console.log(query);
		console.log(params);
	}
});
