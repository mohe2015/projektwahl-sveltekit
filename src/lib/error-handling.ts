// SPDX-License-Identifier: AGPL-3.0-or-later
// SPDX-FileCopyrightText: 2021 Moritz Hedtke <Moritz.Hedtke@t-online.de>

import type { Result } from "./result";

export const myFetch = async <T>(
	url: string,
	options: RequestInit | undefined
): Promise<Result<T>> => {
	try {
		const response = await fetch(url, options);
		if (!response.ok) {
			try {
				const additionalInfo = await response.text();
				return {
					result: "failure",
					failure: {
						network: `Failed to request ${url}: ${response.status} ${response.statusText}\nAdditional information: ${additionalInfo}`
					}
				};
			} catch (error: unknown) {
				return {
					result: "failure",
					failure: {
						network: `Failed to request ${url}: ${response.status} ${response.statusText}`
					}
				};
			}
		}
		const result = (await response.json()) as Result<T>;
		return result;
	} catch (error) {
		console.error(error);
		if (error instanceof TypeError) {
			return {
				result: "failure",
				failure: {
					network: `Failed to request ${url}: ${error.message}\nAdditional information: ${
						error.stack ?? 'none'
					}`
				}
			};
		} else {
			return {
				result: "failure",
				failure: {
					network: `Failed to request ${url}: Unknown error see console.`
				}
			};
		}
	}
};