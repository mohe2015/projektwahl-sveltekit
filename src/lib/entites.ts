// SPDX-License-Identifier: AGPL-3.0-or-later
// SPDX-FileCopyrightText: 2021 Moritz Hedtke <Moritz.Hedtke@t-online.de>
import { browser } from '$app/env';
import type { Load } from '@sveltejs/kit/types';
import { HTTPError } from './authorization';

export type BaseEntityType = {
	id: number;
	[x: string]: string | number | null;
};

export type EntityResponseBody = {
	entities: Array<BaseEntityType>;
	previousCursor: BaseEntityType | null;
	nextCursor: BaseEntityType | null;
};

export type BaseQueryType = {
	'sorting[]': string[];
	pagination_limit: string;
	[x: string]: string | string[];
};
