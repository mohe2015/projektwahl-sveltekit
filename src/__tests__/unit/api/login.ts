// SPDX-License-Identifier: AGPL-3.0-or-later
// SPDX-FileCopyrightText: 2021 Moritz Hedtke <Moritz.Hedtke@t-online.de>
import 'jest';
import type { Login } from 'src/routes/login/index.json';
import { failedFetch, successfulFetch } from '../../../lib/test_utils';

test('invalid login returns error', async () => {
	const result = await failedFetch<{ [key: string]: string }>('http://localhost:3000/login.json', {
		method: 'POST',
		headers: {
			'Content-Type': 'application/json',
			'x-csrf-protection': 'projektwahl'
		},
		body: JSON.stringify({
			name: 'test',
			password: 'test'
		})
	});
	expect(result).toMatchSnapshot();
});

test('successful login returns cookie', async () => {
	const result = await successfulFetch<Login>('http://localhost:3000/login.json', {
		method: 'POST',
		headers: {
			'Content-Type': 'application/json',
			'x-csrf-protection': 'projektwahl'
		},
		body: JSON.stringify({
			name: 'admin',
			password: 'changeme'
		})
	});
	expect(result.session.session_id).toHaveLength(36);
});
