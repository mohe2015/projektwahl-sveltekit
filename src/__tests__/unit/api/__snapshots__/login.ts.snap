// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invalid login returns error 1`] = `
Object {
  "errors": Object {
    "name": "Nutzer existiert nicht!",
  },
}
`;
